src/coreclr/vm/dllimportcallback.cpp

```diff
@@ -210,6 +210,33 @@ VOID CallbackOnCollectedDelegate(UMEntryThunkData* pEntryThunkData)
     EEPOLICY_HANDLE_FATAL_ERROR_WITH_MESSAGE(COR_E_FAILFAST, message.GetUnicode());
 }
 
+#ifdef FEATURE_INTERPRETER
+PLATFORM_THREAD_LOCAL UMEntryThunkData * t_MostRecentUMEntryThunkData;
+
+UMEntryThunkData * GetMostRecentUMEntryThunkData()
+{
+    UMEntryThunkData * result = t_MostRecentUMEntryThunkData;
+    t_MostRecentUMEntryThunkData = nullptr;
+    return result;
+}
+#endif
+
+static void PatchUMEntryPrecodeIfNecessary(UMEntryThunkData * pUMEntryThunkData, PCODE callTarget)
+{
+#ifdef FEATURE_INTERPRETER
+    // HACK: We currently rely on the thread-local stash in TheUMEntryPrestubWorker for proper operation
+    //  of reverse p/invoke IL stubs in the interpreter, so we can't patch the precode in this scenario.
+    // Instead, every reverse p/invoke will go through TheUMEntryPrestubWorker.
+    MethodDesc *pMD = NonVirtualEntry2MethodDesc(callTarget);
+    if (pMD && pMD->GetInterpreterCode())
+        return;
+#endif
+
+    // NOTE: This may be run concurrently by multiple threads, but it should be safe since we will just be
+    //  setting the target of the precode to the same thing multiple times in a row.
+    pUMEntryThunkData->PatchPrecode();
+}
+
 PCODE TheUMEntryPrestubWorker(UMEntryThunkData * pUMEntryThunkData)
 {
     STATIC_CONTRACT_THROWS;
```

Comment:
```suggestion
    STATIC_CONTRACT_THROWS;
    
#ifdef FEATURE_INTERPRETER
     PCODE pInterpreterTarget = pUMEntryThunkData->GetInterpreterTarget()
     if (pInterpreterTarget != NULL)
     {
         t_MostRecentUMEntryThunkData = pUMEntryThunkData;
         return pInterpreterTarget ;
     }
#endif // FEATURE_INTERPRETER
 
```
This should be right after `STATIC_CONTRACT_MODE_PREEMPTIVE` but github does not allow me to attach the comment there.